### PR TITLE
[Merged by Bors] - Added code to perform full cluster stop when changing the deployed NiFi version

### LINKED ISSUE #238: NiFi upgrade doesn't work
### Affected version

0.5.0

### Current and expected behavior

**Scenario**
A NifiCluster with three nodes was deployed with version 1.13.2 and is up and running.

The NifiCluster CRD is now changed to version 1.15.0.

**Current Behavior**
The StatefulSet is updated with the new image and triggers a rolling restart of the NiFi Pods with the new container image set to NiFi 1.15.0.

However NiFi does not support running a cluster with mixed versions, instead a full stop and restart with the new version is required.

Reference ticket: 
https://issues.apache.org/jira/browse/NIFI-4068?jql=project%20%3D%20NIFI%20AND%20text%20~%20%22rolling%20upgrade%22

Due to this, the new pod never successfully starts and the restart hangs indefinitely, or until the user deletes all pods and they are rewritten with the same version by the StatefulSet.

**Expected Behavior**
The operator should notice that a version changes is happening and trigger a full restart of NiFi.

This is done when 
- [ ] I can update my NiFi CR from one NiFi version to another and have the operator automatically run a full-restart of NiFi to have a fully running _new_ NiFi (new = new version) 
- [ ] The solution to implement this can be reused by other operators for similar scenarios
- [ ] The proposed solution has been discussed in the architecture meeting before it's fully implemented
- [ ] The implementation solution has been documented in the Contributor's Guide

### Possible solution

The operator needs to be able to recognize a version change during reconciliation and then act accordingly to perform a full cluster restart.

Something along the lines of this code (suggested by @teozkr ) might work:

```
if current_sts.spec.template.spec.image != new_image:
  if current_sts.status.replicas > 0:
    # Wait for all current replicas to die
    new_sts.spec.replicas = 0
  else:
    # All old replicas are dead, do the upgrade
    new_sts.spec.replicas = rolegroup.replicas
    new_sts.spec.template.spec.image = new_image
```

### Environment

This should be reproducable independently of the K8s environment.

### PR DESCRIPTION
**Pull request by soenkeliebau**
Currenty we cannot upgrade NiFi without manual intervention, as NiFi can't run with mixed versions in a cluster and StateFulSets do not have an upgrade policy of "full".

This PR adds tracking of the deployed NiFi version to the status of the NiFi object and evaluates if a reconciliation signifies a change of the deployed version. If this is recognized a full stop of the cluster is performed and only then the updated version rolled out.

fixes #238

Signed-off-by: Sönke Liebau <soenke.liebau@stackable.tech>

# Description

*Please add a description here. This will become the commit message of the merge request later.*

## Review Checklist

- [ ] Code contains useful comments
- [ ] CRD change approved (or not applicable)
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator-binary/src/controller.rs [hunk 1]
```diff
@@ -193,6 +207,70 @@ pub async fn reconcile_nifi(nifi: Arc<NifiCluster>, ctx: Arc<Ctx>) -> Result<Act
...
+            // Retrieve the deployed statefulsets to check on the current status of the restart
+            let deployed_statefulsets = client
+                .list_with_label_selector::<StatefulSet>(Some(namespace), &selector)
+                .await
+                .context(ApplyRoleServiceSnafu)?;
+
+            // Sum target replicas for all statefulsets
+            let target_replicas = deployed_statefulsets
+                .iter()
+                .filter_map(|statefulset| statefulset.spec.clone())
```
Comment (maltesander) [inline rust/operator-binary/src/controller.rs:234]:
```suggestion
                .filter_map(|statefulset| statefulset.spec.as_ref())
```

File: rust/operator-binary/src/controller.rs [hunk 2]
```diff
@@ -193,6 +207,70 @@ pub async fn reconcile_nifi(nifi: Arc<NifiCluster>, ctx: Arc<Ctx>) -> Result<Act
     tracing::info!("Checking for sensitive key configuration");
     check_or_generate_sensitive_key(client, &nifi).await?;
 
+    // Handle full restarts for a version change
+    let version_change = if let Some(deployed_version) =
+        &nifi.status.clone().unwrap_or_default().deployed_version
```
Comment (maltesander) [inline rust/operator-binary/src/controller.rs:212]: Not prettier but a clone less:
```suggestion
    let version_change = if let Some(deployed_version) = nifi
        .status
        .as_ref()
        .and_then(|status| status.deployed_version.as_ref())
```

File: rust/operator-binary/src/controller.rs [hunk 3]
```diff
@@ -193,6 +207,70 @@ pub async fn reconcile_nifi(nifi: Arc<NifiCluster>, ctx: Arc<Ctx>) -> Result<Act
...
+            let target_replicas = deployed_statefulsets
+                .iter()
+                .filter_map(|statefulset| statefulset.spec.clone())
+                .filter_map(|spec| spec.replicas)
+                .sum::<i32>();
+
+            // Sum current ready replicas for all statefulsets
+            let current_replicas = deployed_statefulsets
+                .iter()
+                .filter_map(|statefulset| statefulset.status.clone())
```
Comment (maltesander) [inline rust/operator-binary/src/controller.rs:241]:
```suggestion
                .filter_map(|statefulset| statefulset.status.as_ref())
```

File: tests/templates/kuttl/upgrade/00-assert.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: kuttl.dev/v1beta1
+kind: TestAssert
+metadata:
+  name: install-zk
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/00-assert.yaml:5]: As i learned recently this doesnt do anything: https://kuttl.dev/docs/testing/reference.html#testassert

**Review by maltesander — CHANGES_REQUESTED** [review]
Tested and works fine (only one zookeeper version)!
```
--- PASS: kuttl (348.76s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.8.0-stackable0.7.0_nifi_old-1.15.0-stackable0.4.0_nifi_new-1.16.3-stackable0.1.0 (164.70s)
        --- PASS: kuttl/harness/upgrade_zookeeper-3.8.0-stackable0.7.0_nifi_old-1.16.0-stackable0.1.0_nifi_new-1.16.3-stackable0.1.0 (182.84s)
PASS
```

Some issues/questions:
1) The (old) reporting task job is not getting deleted (missing in orphaned resources) after update
2) Does it make sense to use one of the python tests (e.g. test_nifi.py to check for registered nodes) in the upgrade tests?
3) Do we need a flow-file/data test? In terms of are the versions and their data really compatible? We do that e.g. in Kafka.
4) Does it make sense to emit the log messages as events ("Cluster has been stopped for a restart, will scale back up.")?

**Comment by soenkeliebau** [comment]
> 1. The (old) reporting task job is not getting deleted (missing in orphaned resources) after update
We need to have a look at that, yes. I noticed this as well, but figured this will mostly be an issue during testing, when we trigger the upgrade before the job for the old version is finished.
For the tests we could solve it "preliminarily" by adding a condition to wait for the job to succeed in the install NiFi step I think..

> 2. Does it make sense to use one of the python tests (e.g. test_nifi.py to check for registered nodes) in the upgrade tests?
It absolutely does, I'll look into this when I get back. Or if someone else wants to pick this up in the meantime, please do feel free to!

> 3. Do we need a flow-file/data test? In terms of are the versions and their data really compatible? We do that e.g. in Kafka.
We should have that, yes. 

> 4. Does it make sense to emit the log messages as events ("Cluster has been stopped for a restart, will scale back up.")?
It does!

File: tests/templates/kuttl/upgrade/00-assert.yaml (hunk 4 above)
Comment (soenkeliebau) [reply tests/templates/kuttl/upgrade/00-assert.yaml:5]: You meant the `name` doesn't do anything?

**Review by razvan — APPROVED** [review]

File: rust/operator-binary/src/controller.rs [hunk 5]
```diff
@@ -193,6 +207,72 @@ pub async fn reconcile_nifi(nifi: Arc<NifiCluster>, ctx: Arc<Ctx>) -> Result<Act
...
+            // Retrieve the deployed statefulsets to check on the current status of the restart
+            let deployed_statefulsets = client
+                .list_with_label_selector::<StatefulSet>(Some(namespace), &selector)
+                .await
+                .context(ApplyRoleServiceSnafu)?;
+
+            // Sum target replicas for all statefulsets
+            let target_replicas = deployed_statefulsets
+                .iter()
+                .filter_map(|statefulset| statefulset.spec.clone())
```
Comment (maltesander) [inline rust/operator-binary/src/controller.rs:236]:
```suggestion
                .filter_map(|statefulset| statefulset.spec.as_ref())
```

File: tests/templates/kuttl/upgrade/00-install-zk.yaml.j2 [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+kind: ZookeeperCluster
+metadata:
+  name: test-zk
+spec:
+  servers:
+    roleGroups:
+      default:
+        replicas: 1
+        config:
+          myidOffset: 10
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/00-install-zk.yaml.j2:12]: This is not required for a 1 role group cluster.

File: tests/templates/kuttl/upgrade/01-install-nifi.yaml.j2 [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        singleUser:
+          adminCredentialsSecret: nifi-admin-credentials-simple
+    sensitiveProperties:
+      keySecret: nifi-sensitive-property-key
+  nodes:
+    roleGroups:
+      default:
+        selector:
+          matchLabels:
+            kubernetes.io/os: linux
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/01-install-nifi.yaml.j2:36]: Could be removed.

File: tests/templates/kuttl/upgrade/02-assert.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: kuttl.dev/v1beta1
+kind: TestAssert
+metadata:
+  name: install-test-nifi
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/02-assert.yaml:5]: Could be removed.

File: tests/templates/kuttl/upgrade/03-assert.yaml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+---
+apiVersion: kuttl.dev/v1beta1
+kind: TestAssert
+metadata:
+  name: test-nifi
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/03-assert.yaml:5]: Could be removed.

File: tests/templates/kuttl/upgrade/04-upgrade-nifi.j2 [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        singleUser:
+          adminCredentialsSecret: nifi-admin-credentials-simple
+    sensitiveProperties:
+      keySecret: nifi-sensitive-property-key
+  nodes:
+    roleGroups:
+      default:
+        selector:
+          matchLabels:
+            kubernetes.io/os: linux
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/04-upgrade-nifi.j2:21]: Could be removed.

File: tests/templates/kuttl/upgrade/07-prepare-test-nifi.yaml [hunk 11]
```diff
@@ -0,0 +1 @@
+---
```
Comment (maltesander) [inline tests/templates/kuttl/upgrade/07-prepare-test-nifi.yaml:1]: Remove the file? Should work with only the 07-assert.yaml as well.

**Review by maltesander — CHANGES_REQUESTED** [review]
Just minor things and a question:
Do we want to try loading a flow template / data in general? 

@sbernauer used a job to ingest the flow: https://github.com/stackabletech/stackablectl/blob/main/demos/nifi-kafka-druid-water-level-data/create-nifi-ingestion-job.yaml and https://github.com/stackabletech/stackablectl/blob/main/demos/nifi-kafka-druid-water-level-data/IngestWaterLevelsToKafka.xml. 
We could check in the up/downgrade that this is still available / working after the restart?

File: tests/templates/kuttl/upgrade/00-install-zk.yaml.j2 (hunk 6 above)
Comment (razvan) [reply tests/templates/kuttl/upgrade/00-install-zk.yaml.j2]: done

**Comment by razvan** [comment]
CI test: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/nifi-operator-it-custom/8/

**Review by maltesander — APPROVED** [review]
Very nice! LGTM.

**Comment by razvan** [comment]
bors merge

**Pull request closed without merging**